connor/traderModule.go

```diff
@@ -0,0 +1,539 @@
+		return 0, err
+	}
+	if actOrder.GetId() != nil && actOrder.GetPrice() != nil {
+		reBuyHash := buyMgHash + buyMgHash*step
+		if err := t.c.db.SaveOrderIntoDB(&database.OrderDb{
+			OrderID:         actOrder.GetId().Unwrap().Int64(),
+			Price:           actOrder.GetPrice().Unwrap().Int64(),
+			Hashrate:        actOrder.GetBenchmarks().GPUEthHashrate(),
+			StartTime:       time.Now(),
+			ButterflyEffect: int32(actOrder.GetOrderStatus()),
+			ActualStep:      reBuyHash,
+		}); err != nil {
+			return 0, fmt.Errorf("cannot save order to database: %v", err)
+		}
+		log.Printf("Order created ==> ID: %v, Price: %v $, Hashrate: %v H/s \r\n",
+			actOrder.GetId().Unwrap().Int64(),
+			t.PriceToString(actOrder.GetPrice().Unwrap()),
+			actOrder.GetBenchmarks().GPUEthHashrate())
+		return reBuyHash, nil
+	}
+	return buyMgHash, nil
+}
+
+func (t *TraderModule) GetProfitForTokenBySymbol(tokens []*TokenMainData, symbol string) (float64, error) {
+	for _, t := range tokens {
+		if t.Symbol == symbol {
+			return t.ProfitPerMonthUsd, nil
+		}
+	}
+	return 0, fmt.Errorf("Cannot get price from token! ")
+}
+
+func (t *TraderModule) GetPriceForTokenPerSec(token watchers.TokenWatcher, symbol string) (float64, float64, error) {
+	tokens, err := t.profit.CollectTokensMiningProfit(token)
+	if err != nil {
+		return 0, 0, fmt.Errorf("Cannot calculate token prices: %v\r\n", err)
+	}
+	pricePerMonthUSD, err := t.GetProfitForTokenBySymbol(tokens, symbol)
+	if err != nil {
+		return 0, 0, fmt.Errorf("Cannot get profit for tokens: %v\r\n", err)
+	}
+	pricePerSec := pricePerMonthUSD / (secsPerDay * daysPerMonth)
+	return pricePerMonthUSD, pricePerSec, nil
+}
+
+func (t *TraderModule) ReinvoiceOrder(ctx context.Context, cfg *Config, price *sonm.Price, bench map[string]uint64, tag string) error {
+	order, err := t.c.Market.CreateOrder(ctx, &sonm.BidOrder{
+		Duration: &sonm.Duration{Nanoseconds: 0},
+		Price:    price,
+		Tag:      tag,
+		Identity: cfg.OtherParameters.IdentityForBid,
+		Resources: &sonm.BidResources{
+			Network: &sonm.BidNetwork{
+				Overlay:  false,
+				Outbound: true,
+				Incoming: false,
+			},
+			Benchmarks: bench,
+		},
+	})
+	if err != nil {
+		fmt.Printf("Cannot created Lucky Order: %v\r\n", err)
+		return err
+	}
+	if err := t.c.db.SaveOrderIntoDB(&database.OrderDb{
+		OrderID:         order.GetId().Unwrap().Int64(),
+		Price:           order.GetPrice().Unwrap().Int64(),
+		Hashrate:        order.GetBenchmarks().GPUEthHashrate(),
+		StartTime:       time.Now(),
+		ButterflyEffect: int32(OrderStatusReinvoice),
+		ActualStep:      0,
+	}); err != nil {
+		return fmt.Errorf("cannot save reinvoice order %s to DB: %v \r\n", order.GetId().Unwrap().String(), err)
+	}
+	log.Printf("REINVOICE Order ===> %v created (descendant: %v), price: %v, hashrate: %v",
+		order.GetId(), tag, order.GetPrice(), order.GetBenchmarks().GPUEthHashrate())
+	return nil
+}
+
+func (t *TraderModule) CmpChangeOfPrice(change float64, def float64) (int32, error) {
```

Comment:
what the weird magic?

Reply:
This function compares the price change.

Reply:
yeap, I can read some English words that are in a function name. I mean a magic inside the method.